JavaScript/chapter02/p08_loop_detection/avc278.js

```diff
@@ -0,0 +1,100 @@
+// Loop Detection: Given a circular linked list, implement an algorithm that returns the node at the beginning of the
+// loop.
+// Example: A -> B -> C -> D
+//                    ^    |
+//                    |    v
+//                    | <- E
+// Output: C
+const assert = require("assert");
+const { LinkedListNode } = require("../../lib/avc278/linkedlist");
+
+/**
+ *
+ * @param   {LinkedListNode} list input linked list in which to find loop
+ * @return  {LinkedListNode}      the node in the linked list that starts the loop
+ *
+ * Overall in this problem, we only keep track of pointers and a boolean, so we use O(1) additional space.
+ * In terms of runtime complexity, we only iterate through the entire linked list one full time to find out where the
+ * slow and fast runners intersect. After this, we start iterating through the linked list again, which in the worst
+ * case, would result in iterating through the entire list again, but still only requires O(N) time, where N is the
+ * size of the linked list.
+ * Runtime: O(N)
+ * Space:   O(1)
+ *
+ */
+const loopDetection = (list) => {
+  let slow = list;
+  let fast = list;
+  let intersect = false;
+
+  while (!intersect) {
+    slow = slow.next;
+    fast = fast.next.next;
+
+    if (slow !== fast) continue;
+    intersect = true;
+  }
+  slow = list;
+
+  while (slow !== fast) {
+    slow = slow.next;
+    fast = fast.next;
+  }
```

Comment:
Can there be an explanation as to why this (and the loop in L30) work? Not that I don't believe the code is correct, but this is "magical" to most folks.

Reply:
Yeah, good point. I'll make a short write-up and update the solution file. This was definitely magic to me the first time I stumbled upon this solution a few weeks ago.